cpp/src/sender.cpp

```diff
@@ -64,10 +64,18 @@ namespace {
 uint64_t tag_counter = 0;
```

Comment:
AFAIK thread safety is required, to guarantee unique tags. Or a different way of generating them should be found. Something for later, to investigate and fix.

Reply:
Should I file the bug for this or does it already exists?

Reply:
Thanks, I will take a look.

Reply:
@jiridanek Is the C++ binding sender actually considered thread-safe? I definitely didnt believe it to be, I think none of them are, and the C stuff its calling underneath certainly isnt. If its not then the tag generation doesnt seem a concern. The [pre-existing] approach used clearly isnt thread safe, multiple threads cant safely pre-/post-increment a simple shared value without some protection, but thats fine if it isnt intended or needed to be. The link usage in the send method also appears unprotected to my [admittedly C++ oblivious] eyes, which would mean its all inherently unsafe anyway, and duplicate tags is likely on the lesser end of issues that could come up in there.

Reply:
Thread safe is a matter of degree, of course. The situation I am thinking about is program with two threads, each running its own container and not interacting in any way with the other thread. I believed this is something that Proton supports or at least strives to support. In this case, because `uint64_t tag_counter = 0;` is a global variable, the two threads will step over each other's toes anyways and will send duplicate tags.

My suggestion for Raki here was to create a Jira ticket for it. Do you think what's happening is completely expected and raising Jira is not warranted? I fully agree that the issue has low priority. It has been this way for ages and noone ever complained.

Reply:
> In this case, because `uint64_t tag_counter = 0;` is a global variable, the two threads will step over each other's toes anyways and will send duplicate tags.

Ah ok. Yes they would in that case. I wasnt realising it was global, I thought it was a sender-specific variable in the senders case (I told you my eyes are C++ oblivious). If its global thats clearly a bug that should be raised, then it can be fixed in isolation.

Reply:
@DreamPearl please fill a jira then, and close this thread. Thanks!

Reply:
Okay. Done!

Reply:
For later reference, the new JIRA is https://issues.apache.org/jira/browse/PROTON-2377